### ISSUE TITLE
Add Infection Mutation Testing as Library 

### ISSUE THREAD

**Issue by larsroettig** [magento/magento2#18478]
### Description
As Developer, I would use Mutation Testing for my Magento Implementations.

### Expected behavior
-  Injection is integrated https://infection.github.io/guide/installation.html

### Benefits
- It is a powerful approach to attain high coverage of the source program.
- This testing is capable comprehensively testing the mutant program.
- Mutation testing brings a good level of error detection to the software developer.
- This method uncovers ambiguities in the source code, and has the capacity to detect all the faults in the program.
- Customers are benefited from this testing by getting most reliable and stable system.

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@larsroettig do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by paliarush** [comment]
Makes sense to try it out to see how well it works for the Magento codebase.
In scope of prototype we should evaluate:
1. For which kinds of tests mutation testing makes sense (integration, MFTF, web API, etc)
2. Tests execution time implications
3. Considering the size of Magento codebase, we might only need to "infest" select modules while running tests for these modules. We need to think about other possible optimizations
4. Is it possible to get repeatable results? How do we verify that the fixed issue? 
5. Are we going to integrate it into PR delivery process or run once in a while manually?

**Comment by theofidry** [comment]
Given the size of the Magento codebase, I would really target unit tests since otherwise the waiting time may be too long. Using it incrementally, i.e. just on the files changed, would also make it faster.

>Is it possible to get repeatable results? How do we verify that the fixed issue?

If your tests are deterministic yes it is: mutants are generated and applied in a deterministic way. You can verify the issue is fixed by running the tests again, the process usually is:

- run mutation testing, get the report with the escaped mutants
- apply an escaped mutant manually and add a test case to capture it (i.e. add a failing test case) and then revert the escaped mutant
- you can go a step further and try to remove some tests see if they are redundant or not (easier to apply on new elements of the codebase)

>Are we going to integrate it into PR delivery process or run once in a while manually?

Running it once would help to raise the quality one time which is not bad but I would consider adding it to the "delivery process" process if by that you mean PRs when adding or refactoring elements.

**Comment by theofidry** [comment]
Looks like you can't ping other repositories teams... So:

/cc @sanmai, @sidz, @borNfreee, @BackEndTea

**Comment by larsroettig** [comment]
@paliarush maybe it is a good idea to start with only PHPUnit Test and use for Travis only the changes from the PR( https://backendtea.github.io/2018/06/10/test-your-tests-are-testing/#usage).

Next Step is for me to have a night build or something like that to find bugs in the core modules. 

What do you mean it this possible to that?

**Comment by sanmai** [comment]
Infection is not just a theoretical tool, but a very practical. Best would be to start with a PoC run, to show practical test deficiencies. Or even bugs: sometime, but not often, you can find that that's not the test broken, but there's an actual bug.
